facebook/jest
# jest-snapshot: Remove report method and throw matcher errors

## Summary

1. Remove `report` method from returned object (added in #2130 for AVA, but ironically was not called then, nor called now by any dependent of `jest-snapshot` package)

2. Throw matcher errors when assertion fails because invalid arguments:

    * Snapshot assertion must not have **not**
    * Expected properties must be an object
    * Inline snapshot must be a string
    * Received value must be a function

3. Do not eagerly precompute `option` because it is not used if an assertion passes

4. Call new `matcherHintFromConfig` function after matcher arguments have been validated:

    * Snapshot state must be initialized (error)
    * Received function did not throw (error, but see Residue below)
    * New snapshot was **not written** (message)
    
5. Call renamed `printSnapshotName` function whenever the `count` is known:

    * properties did not match (improved)
    * snapshot did not match (baseline)
    * New snapshot was **not written** (improved)

6. Improve type annotation of `propertiesOrHint` and `propertiesOrSnapshot` **Question**: do y’all recommend something better than `object` type?

7. Make third argument optional in `matcherErrorMessage` from `jest-matcher-utils`

Rename `print.ts` as `printSnapshot.ts`

**Residue**

* Does `toThrowErrorMatchingInlineSnapshot` need to call `stripAddedIndentation`?
* When received function did not throw, call `fail` instead of throw error?

To minimize breaking changes in Jest 25, wait until next version:

* Matcher error: Snapshot hint must be a string
* Add `|| hint` to `fullTestName` and move it to `utils.ts`

## Test plan

Rename `printDiffOrStringified.test.*` as `printSnapshot.test.*`

* Update 30 snapshot names after enclose existing tests in `describe` block
* Add 19 tests to see results and increase code coverage of `index.ts` file

| e2e test | change |
| ---: | :--- |
| `toMatchSnapshot` | update 1 test |
| `toThrowErrorMatchingInlineSnapshot` | update 1 test |
| `toThrowErrorMatchingSnapshot` | update 1 test |

See also pictures in following comments: baseline is at left and improved is at right
